billing/models.py

```diff
@@ -1,3 +1,103 @@
+import datetime
+from typing import Union
+
+from django.core.exceptions import ValidationError
 from django.db import models
+from django.utils import timezone
+
+from billing.common.constants import BillingConstants
+from billing.common.constants import CurrencyConstants
+from billing.common.strings import AccountingMethodText
+from billing.common.strings import BillingDetailsAbstractModelText
+from billing.common.strings import BillingValidationErrorText
+from managers.models import Project
+from users.common.fields import ChoiceEnum
+from users.models import CustomUser
+from utils.models import BaseModel
+
+
+class BillingDetailsAbstractQuerySet(models.QuerySet):
+    pass
+
+
+class BillingDetailsAbstract(BaseModel):
+    class Meta:
+        abstract = True
+
+    class Currency(ChoiceEnum):
+        PLN = CurrencyConstants.CurrencyDict["PLN"]["Code"]
+        USD = CurrencyConstants.CurrencyDict["USD"]["Code"]
+        EUR = CurrencyConstants.CurrencyDict["EUR"]["Code"]
+
+    class AccountingMethod(ChoiceEnum):
+        HOURLY = AccountingMethodText.HOURLY.value
+        MONTHLY = AccountingMethodText.MONTHLY.value
+
+    user = models.ForeignKey(CustomUser, on_delete=models.PROTECT, related_name="%(class)s")
+    value = models.PositiveSmallIntegerField(BillingDetailsAbstractModelText.VALUE.value, blank=False, null=False)
+    currency = models.CharField(
+        max_length=BillingConstants.CURRENCY_MAX_LENGTH.value, choices=Currency.choices(), default=Currency.PLN
+    )
+    accounting_method = models.CharField(
+        max_length=BillingConstants.ACCOUNTING_METHOD_MAX_LENGTH.value,
+        choices=AccountingMethod.choices(),
+        default=AccountingMethod.HOURLY.name,
+    )
+    from_date = models.DateTimeField(BillingDetailsAbstractModelText.FROM_DATE.value, blank=False, null=False)
+    to_date = models.DateTimeField(BillingDetailsAbstractModelText.TO_DATE.value, blank=True, null=True)
+
+    objects = BillingDetailsAbstractQuerySet.as_manager()
+
+    def clean(self) -> None:
+        super().clean()
+
+        if self.to_date is not None and self.to_date <= self.from_date:
+            raise ValidationError(
+                message=BillingValidationErrorText.VALIDATION_ERROR_FROM_DATE_LATER_THAN_TO_DATE.value
+            )
+        if self.to_date is None and not self.user.is_active:
+            raise ValidationError(
+                message=BillingValidationErrorText.VALIDATION_ERROR_TO_DATE_EMPTY_FOR_INACTIVE_USER.value
+            )
```

Comment:
Why check for `self.to_date`? What if user inactive but `self.to_date` is given?